crates/bevy_ecs/src/system/function_system.rs

```diff
@@ -790,7 +792,14 @@ where
 
         for archetype in &archetypes[old_generation..] {
             // SAFETY: The assertion above ensures that the param_state was initialized from `world`.
-            unsafe { F::Param::new_archetype(&mut state.param, archetype, &mut self.system_meta) };
+            unsafe {
+                F::Param::new_archetype(
+                    &state.param,
+                    archetype,
+                    &mut self.system_meta,
+                    world.world(),
```

Comment:
The calls to `world()` here and above aren't sound.  `update_archetype_component_access` gets called by the multi-threaded executor while other systems are running, so we don't have permission to access the world immutably.

Hmm, this PR removes the need for this to update the `QueryState`, but it still needs to be here to update the `archetype_component_access`.  If we did #16885 first, then we could get rid of this call entirely!

Reply:
Yeah you're right, I'm in favor of landing #16885 first, even though with QueryState as immutable component we're half safe now haha